fix: gf fails on foo/bar.txt:line:column filepath (Windows)

lsvmello opened the pull request:
First time contributing!

The solution has to remove colon (:) from `isfname` on Windows, so I marked the commit with the breaking change flag.
I treated `:\` it as a special case just like `:/`.

# Problem

On Windows, "gf" fails on a filepath that has a line:column suffix. Example:

    E447: Can't find file "src/app/core/services/identity/identity.service.ts:64:23"

# Solution

Colon is not a valid filename character on Windows (except for the drive-letter), so including ":" in 'isfname' is wrong.

Fixes #25160
Related #24032

lsvmello committed:
fix: gf fails on "foo/bar.txt:1:2" on Windows

Problem:
On Windows, "gf" fails on a filepath that has a line:column suffix.
Example:

    E447: Can't find file "src/app/core/services/identity/identity.service.ts:64:23"

Solution:
- Remove ":" from 'isfname' on Windows. Colon is not a valid filename
  character (except for the drive-letter).
- Handle drive letters specially in file_name_in_line().

Fixes #25160

zeertzjq commented:
Try changing 'isfname' to the old value in `test/old/testdir/setup.vim` instead.

zeertzjq commented:
> > Try changing 'isfname' to the old value in `test/old/testdir/setup.vim` instead.
> 
> Instead of updating the `test/old/testdir/test_regexp_utf8.vim`?

Yes.

lsvmello commented:
Done, but it seems other things have crashed now.

justinmk commented on src/nvim/window.c:
```diff
@@ -6945,15 +6945,18 @@ char *file_name_in_line(char *line, int col, int options, int count, char *rel_f
   // Search forward for the last char of the file name.
-  // Also allow ":/" when ':' is not in 'isfname'.
-  len = 0;
+  // Also allow ":/" and ":\" when ':' is not in 'isfname'.
```
"/" is usually, internally, understood to mean "any valid path separate" in most contexts. Is the comment change actually necessary?

lsvmello replied:
I think the comment is there to clarify about the logic that considers ':' even if not in 'isfname'. I thought that it might be useful to add that ':\' also is checked, but if we want to be very explicit all ':\', ':/', ':?' and ':#' are actually considered on `path_has_drive_letter` function.

zeertzjq replied:
Note that this comment was originally referring to URLs.

lsvmello replied:
Yes but the logic was exclusively for URLs, now it also considers drives

justinmk commented on test/old/testdir/setup.vim:
```diff
@@ -30,6 +30,10 @@ if exists('s:did_load')
   set switchbuf=
+  if has('win32')
+    " set 'isfname' as old win32 default value
+    set isfname=@,48-57,/,\,.,-,_,+,,,#,$,%,{,},[,],:,@-@,!,~,=
```
```suggestion
    set isfname+=:
```

justinmk commented on runtime/doc/news.txt:
```diff
@@ -15,6 +15,8 @@ BREAKING CHANGES                                                *news-breaking*
 
 The following changes may require adaptations in user config or plugins.
 
+• colon (:) has been removed from |'isfname'| on Windows.
```
The news item should give a hint about the meaning of the change, e.g. "File path drive letters are now detected on Windows without needing to explicitly put ':' in 'isfname'." 

Also need to update `:h nvim-defaults`.

justinmk commented on runtime/doc/news.txt:
```diff
@@ -15,6 +15,9 @@ BREAKING CHANGES                                                *news-breaking*
 
 The following changes may require adaptations in user config or plugins.
 
+• File path drive letters are now detected on Windows without needing to
+  explicitly put ':' in 'isfname'.
```
```suggestion
• Windows file path drive letters are now detected even if ":" is not in 'isfname'.
   And 'isfname' now omits ":" by default on Windows.
```

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -53,6 +53,7 @@ Defaults					            *nvim-defaults*
 - 'hlsearch' is enabled
 - 'include' defaults to "". The C ftplugin sets it to "^\\s*#\\s*include"
 - 'incsearch' is enabled
+- 'isfname' does not include ':' on Windows
```
```suggestion
- 'isfname' does not include ":" on Windows. Include ":" in 'isfname' to treat it as part of a filename following the drive letter. 
```

justinmk commented on src/nvim/window.c:
```diff
@@ -6934,11 +6934,13 @@ char *file_name_in_line(char *line, int col, int options, int count, char *rel_f
 
   // Search backward for first char of the file name.
-  // Go one char back to ":" before "//" even when ':' is not in 'isfname'.
+  // Go one char back to ":" before "//" or
+  // two chars back to the alphanumeric befor ":\"
+  // even when ':' is not in 'isfname'.
```
```suggestion
  // Go one char back to ":" before "//" or two chars back before ":/" even if ":" is not in 'isfname'.
```

justinmk reviewed (commented):
This is looking better, but needs an explicit test to clearly show what's expected. Can be easily tested using `expand('<cfile>')`. 

- This is a good place for the test to live: https://github.com/neovim/neovim/blob/5ea5812714bcbeb6a5438c80d410984ad4fbb922/test/functional/core/path_spec.lua#L60
- Please check that the test correctly _fails_ if the code changes are reverted.

You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#running-tests) like this:

    TEST_FILE=test/functional/core/path_spec.lua make functionaltest

lsvmello commented:
> * Please check that the test correctly _fails_ if the code changes are reverted.

I have confirmed that the test fails for Windows if the changes are reverted back, but they still pass on the linux case because the path is identified as an URL.

Is is better to:
- skip the test on Linux?
- `set isfname+=\` on Linux to test `C:\test_file`?

justinmk commented on test/functional/core/path_spec.lua:
```diff
@@ -13,46 +13,46 @@ describe('path collapse', function()
   it('with ./../ and different starting directory #7117', function()
     command('cd test')
-    command('edit '..join_path('.', '..', targetdir, 'tty-test.c'))
+    command('edit ' .. join_path('.', '..', targetdir, 'tty-test.c'))
```
?

justinmk commented on test/functional/core/path_spec.lua:
```diff
@@ -67,4 +67,11 @@ describe('file search', function()
+
+  it('considers Windows drives even without ":" in `isfname`', function()
+    command('set isfname-=:')
+    local file_name = is_os('win') and 'C:\\test_file' or 'C:/test_file'
```
the "/" case should also be tested on Windows

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -53,6 +53,8 @@ Defaults					            *nvim-defaults*
 - 'include' defaults to "". The C ftplugin sets it to "^\\s*#\\s*include"
 - 'incsearch' is enabled
+- 'isfname' does not include ":" on Windows. Include ":" in 'isfname' to treat
+  it as part of a filename following the drive letter.
```
```suggestion
  it as part of a filename anywhere in the name (not only the drive letter).
```

justinmk commented on test/functional/core/path_spec.lua:
```diff
@@ -91,4 +91,80 @@ describe('file search', function()
     eq('filename_with_unicode_ααα', eval('expand("%:t")'))
   end)
+
+  it('matches paths even without ":" in `isfname`', function()
```
```suggestion
  it('matches Windows drive-letter filepaths (without ":" in &isfname)', function()
```

justinmk commented on test/functional/core/path_spec.lua:
```diff
@@ -91,4 +91,80 @@ describe('file search', function()
+    local os_win = is_os('win')
+
+    insert([[c:/d:/foo/bar.txt]])
+    eq([[c:/d:/foo/bar.txt]], eval('expand("<cfile>")'))
```
This doesn't seem right. If ":" is not in 'isfname' why is the 2nd drive letter allowed? I would expect the result to be `d:/foo/bar.txt`

Not a blocker for this PR I guess.

lsvmello replied:
My guess is that this happens because of the `path_is_url` implementation. Since just a `:\` characterizes the path as a valid URL then Neovim now considers the following `:` as a valid character because it is used for port specification.

If this becomes a big deal then we might need to change `path_is_url` to implement a URL that follows the [RFC 7230](https://datatracker.ietf.org/doc/html/rfc7230#section-2.7).

justinmk replied:
`path_is_url` should definitely be improved, an easy improvement would be to treat single-char schemes as drive letters (i.e. "not a url"). That's mentioned at https://url.spec.whatwg.org/#start-with-a-windows-drive-letter

ntdiary replied:
> `path_is_url` should definitely be improved, an easy improvement would be to treat single-char schemes as drive letters (i.e. "not a url").

Seems we can switch to `path_with_url`, That's what I tried in #41444 (in `file_name_in_line`).
On Windows, the expected result is `c:/d`; on Unix, it could be `c` or the entire string, depending on url detection. As for url detection, might discuss in that PR later. :)

justinmk committed:
refactor: cleanup

justinmk commented:
Pushed an update with some cleanup + extra tests. Will merge after CI.

justinmk commented:
asan found an issue

```
: heap-buffer-overflow on address 0x6020000079cf at pc 0x563e98bd70b6 bp 0x7ffc52efc430 sp 0x7ffc52efbbf8
= READ of size 1 at 0x6020000079cf thread T0
=     #0 0x563e98bd70b5 in strlen (/home/runner/work/neovim/neovim/build/bin/nvim+0xc7b0b5) (BuildId: e40ec79ca4eae53c0155aeb03de88ddd5d29c5d2)
=     #1 0x563e99da4f45 in path_has_drive_letter /home/runner/work/neovim/neovim/src/nvim/path.c:1754:10
=     #2 0x563e9a610d81 in file_name_in_line /home/runner/work/neovim/neovim/src/nvim/window.c:6951:49
=     #3 0x563e9a61066f in file_name_at_cursor /home/runner/work/neovim/neovim/src/nvim/window.c:6917:10
=     #4 0x563e995a08d6 in eval_vars /home/runner/work/neovim/neovim/src/nvim/ex_docmd.c:6903:16
=     #5 0x563e9937c648 in f_expand /home/runner/work/neovim/neovim/src/nvim/eval/funcs.c:1754:20
=     #6 0x563e99357819 in call_internal_func /home/runner/work/neovim/neovim/src/nvim/eval/funcs.c:273:3
=     #7 0x563e994656de in call_func /home/runner/work/neovim/neovim/src/nvim/eval/userfunc.c:1719:15
=     #8 0x563e99461ae6 in get_func_tv /home/runner/work/neovim/neovim/src/nvim/eval/userfunc.c:557:11
=     #9 0x563e992c7c5d in eval_func /home/runner/work/neovim/neovim/src/nvim/eval.c:2281:13
=     #10 0x563e992bf708 in eval7 /home/runner/work/neovim/neovim/src/nvim/eval.c:3208:15
=     #11 0x563e992bbda9 in eval6 /home/runner/work/neovim/neovim/src/nvim/eval.c:2935:7
=     #12 0x563e992b9e8d in eval5 /home/runner/work/neovim/neovim/src/nvim/eval.c:2791:7
=     #13 0x563e992b87e2 in eval4 /home/runner/work/neovim/neovim/src/nvim/eval.c:2666:7
=     #14 0x563e992b758d in eval3 /home/runner/work/neovim/neovim/src/nvim/eval.c:2575:7
=     #15 0x563e9926299d in eval2 /home/runner/work/neovim/neovim/src/nvim/eval.c:2497:7
=     #16 0x563e99250b0c in eval1 /home/runner/work/neovim/neovim/src/nvim/eval.c:2401:7
=     #17 0x563e9924d68a in eval0 /home/runner/work/neovim/neovim/src/nvim/eval.c:2346:9
=     #18 0x563e98f17315 in nvim_eval /home/runner/work/neovim/neovim/src/nvim/api/vimscript.c:170:3
=     #19 0x563e98e7bb5e in handle_nvim_eval /home/runner/work/neovim/neovim/build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:8953:15
=     #20 0x563e99b62f59 in request_event /home/runner/work/neovim/neovim/src/nvim/msgpack_rpc/channel.c:444:19
=     #21 0x563e9a1dcdc9 in state_handle_k_event /home/runner/work/neovim/neovim/src/nvim/state.c:115:7
=     #22 0x563e99bf0718 in nv_event /home/runner/work/neovim/neovim/src/nvim/normal.c:6623:3
=     #23 0x563e99bb4069 in normal_execute /home/runner/work/neovim/neovim/src/nvim/normal.c:1227:3
=     #24 0x563e9a1dcb8d in state_enter /home/runner/work/neovim/neovim/src/nvim/state.c:97:26
=     #25 0x563e99b9673c in normal_enter /home/runner/work/neovim/neovim/src/nvim/normal.c:516:3
=     #26 0x563e98c82e0d in main /home/runner/work/neovim/neovim/src/nvim/main.c:643:3
=     #27 0x7fcb81429d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 229b7dc509053fe4df5e29e8629911f0c3bc66dd)
=     #28 0x7fcb81429e3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: 229b7dc509053fe4df5e29e8629911f0c3bc66dd)
=     #29 0x563e98bc0f74 in _start (/home/runner/work/neovim/neovim/build/bin/nvim+0xc64f74) (BuildId: e40ec79ca4eae53c0155aeb03de88ddd5d29c5d2)
= 
= 0x6020000079cf is located 1 bytes to the left of 9-byte region [0x6020000079d0,0x6020000079d9)
= allocated by thread T0 here:
=     #0 0x563e98c43dbe in __interceptor_malloc (/home/runner/work/neovim/neovim/build/bin/nvim+0xce7dbe) (BuildId: e40ec79ca4eae53c0155aeb03de88ddd5d29c5d2)
=     #1 0x563e99a7b6f6 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:89:15
=     #2 0x563e99a7b8dc in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:123:15
=     #3 0x563e9901ace2 in ins_str /home/runner/work/neovim/neovim/src/nvim/change.c:801:16
=     #4 0x563e99208133 in insertchar /home/runner/work/neovim/neovim/src/nvim/edit.c:2172:5
=     #5 0x563e9921936c in insert_special /home/runner/work/neovim/neovim/src/nvim/edit.c:1995:5
=     #6 0x563e99218e9f in insert_handle_key /home/runner/work/neovim/neovim/src/nvim/edit.c:1173:7
=     #7 0x563e991f0f1d in insert_execute /home/runner/work/neovim/neovim/src/nvim/edit.c:671:10
=     #8 0x563e9a1dcb8d in state_enter /home/runner/work/neovim/neovim/src/nvim/state.c:97:26
=     #9 0x563e991f8c75 in insert_enter /home/runner/work/neovim/neovim/src/nvim/edit.c:338:5
=     #10 0x563e991ed4e2 in edit /home/runner/work/neovim/neovim/src/nvim/edit.c:1270:3
=     #11 0x563e99bf6007 in invoke_edit /home/runner/work/neovim/neovim/src/nvim/normal.c:6269:7
=     #12 0x563e99bcb665 in nv_edit /home/runner/work/neovim/neovim/src/nvim/normal.c:6246:5
=     #13 0x563e99bb4069 in normal_execute /home/runner/work/neovim/neovim/src/nvim/normal.c:1227:3
=     #14 0x563e9a1dcb8d in state_enter /home/runner/work/neovim/neovim/src/nvim/state.c:97:26
=     #15 0x563e99b9673c in normal_enter /home/runner/work/neovim/neovim/src/nvim/normal.c:516:3
=     #16 0x563e98c82e0d in main /home/runner/work/neovim/neovim/src/nvim/main.c:643:3
=     #17 0x7fcb81429d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 229b7dc509053fe4df5e29e8629911f0c3bc66dd)

justinmk committed:
fix: heap-buffer-overflow in file_name_in_line

    heap-buffer-overflow on address 0x6020000079cf at pc 0x563e98bd70b6 bp 0x7ffc52efc430 sp 0x7ffc52efbbf8
    = READ of size 1 at 0x6020000079cf thread T0
    =     0 0x563e98bd70b5 in strlen (/home/runner/work/neovim/neovim/build/bin/nvim+0xc7b0b5) (BuildId: e40ec79ca4eae53c0155aeb03de88ddd5d29c5d2)
    =     1 0x563e99da4f45 in path_has_drive_letter /home/runner/work/neovim/neovim/src/nvim/path.c:1754:10
    =     2 0x563e9a610d81 in file_name_in_line /home/runner/work/neovim/neovim/src/nvim/window.c:6951:49
    =     3 0x563e9a61066f in file_name_at_cursor /home/runner/work/neovim/neovim/src/nvim/window.c:6917:10
    =     4 0x563e995a08d6 in eval_vars /home/runner/work/neovim/neovim/src/nvim/ex_docmd.c:6903:16
    =     5 0x563e9937c648 in f_expand /home/runner/work/neovim/neovim/src/nvim/eval/funcs.c:1754:20
    =     6 0x563e99357819 in call_internal_func /home/runner/work/neovim/neovim/src/nvim/eval/funcs.c:273:3
    =     7 0x563e994656de in call_func /home/runner/work/neovim/neovim/src/nvim/eval/userfunc.c:1719:15
    =     8 0x563e99461ae6 in get_func_tv /home/runner/work/neovim/neovim/src/nvim/eval/userfunc.c:557:11
    =     9 0x563e992c7c5d in eval_func /home/runner/work/neovim/neovim/src/nvim/eval.c:2281:13
    =     10 0x563e992bf708 in eval7 /home/runner/work/neovim/neovim/src/nvim/eval.c:3208:15
    =     11 0x563e992bbda9 in eval6 /home/runner/work/neovim/neovim/src/nvim/eval.c:2935:7
    =     12 0x563e992b9e8d in eval5 /home/runner/work/neovim/neovim/src/nvim/eval.c:2791:7
    =     13 0x563e992b87e2 in eval4 /home/runner/work/neovim/neovim/src/nvim/eval.c:2666:7
    =     14 0x563e992b758d in eval3 /home/runner/work/neovim/neovim/src/nvim/eval.c:2575:7
    =     15 0x563e9926299d in eval2 /home/runner/work/neovim/neovim/src/nvim/eval.c:2497:7
    =     16 0x563e99250b0c in eval1 /home/runner/work/neovim/neovim/src/nvim/eval.c:2401:7
    =     17 0x563e9924d68a in eval0 /home/runner/work/neovim/neovim/src/nvim/eval.c:2346:9
    =     18 0x563e98f17315 in nvim_eval /home/runner/work/neovim/neovim/src/nvim/api/vimscript.c:170:3
    =     19 0x563e98e7bb5e in handle_nvim_eval /home/runner/work/neovim/neovim/build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:8953:15
    =     20 0x563e99b62f59 in request_event /home/runner/work/neovim/neovim/src/nvim/msgpack_rpc/channel.c:444:19
    =     21 0x563e9a1dcdc9 in state_handle_k_event /home/runner/work/neovim/neovim/src/nvim/state.c:115:7
    =     22 0x563e99bf0718 in nv_event /home/runner/work/neovim/neovim/src/nvim/normal.c:6623:3
    =     23 0x563e99bb4069 in normal_execute /home/runner/work/neovim/neovim/src/nvim/normal.c:1227:3
    =     24 0x563e9a1dcb8d in state_enter /home/runner/work/neovim/neovim/src/nvim/state.c:97:26
    =     25 0x563e99b9673c in normal_enter /home/runner/work/neovim/neovim/src/nvim/normal.c:516:3
    =     26 0x563e98c82e0d in main /home/runner/work/neovim/neovim/src/nvim/main.c:643:3
    =     27 0x7fcb81429d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 229b7dc509053fe4df5e29e8629911f0c3bc66dd)
    =     28 0x7fcb81429e3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: 229b7dc509053fe4df5e29e8629911f0c3bc66dd)
    =     29 0x563e98bc0f74 in _start (/home/runner/work/neovim/neovim/build/bin/nvim+0xc64f74) (BuildId: e40ec79ca4eae53c0155aeb03de88ddd5d29c5d2)
    =
    = 0x6020000079cf is located 1 bytes to the left of 9-byte region [0x6020000079d0,0x6020000079d9)
    = allocated by thread T0 here:
    =     0 0x563e98c43dbe in __interceptor_malloc (/home/runner/work/neovim/neovim/build/bin/nvim+0xce7dbe) (BuildId: e40ec79ca4eae53c0155aeb03de88ddd5d29c5d2)
    =     1 0x563e99a7b6f6 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:89:15
    =     2 0x563e99a7b8dc in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:123:15
    =     3 0x563e9901ace2 in ins_str /home/runner/work/neovim/neovim/src/nvim/change.c:801:16
    =     4 0x563e99208133 in insertchar /home/runner/work/neovim/neovim/src/nvim/edit.c:2172:5
    =     5 0x563e9921936c in insert_special /home/runner/work/neovim/neovim/src/nvim/edit.c:1995:5
    =     6 0x563e99218e9f in insert_handle_key /home/runner/work/neovim/neovim/src/nvim/edit.c:1173:7
    =     7 0x563e991f0f1d in insert_execute /home/runner/work/neovim/neovim/src/nvim/edit.c:671:10
    =     8 0x563e9a1dcb8d in state_enter /home/runner/work/neovim/neovim/src/nvim/state.c:97:26
    =     9 0x563e991f8c75 in insert_enter /home/runner/work/neovim/neovim/src/nvim/edit.c:338:5
    =     10 0x563e991ed4e2 in edit /home/runner/work/neovim/neovim/src/nvim/edit.c:1270:3
    =     11 0x563e99bf6007 in invoke_edit /home/runner/work/neovim/neovim/src/nvim/normal.c:6269:7
    =     12 0x563e99bcb665 in nv_edit /home/runner/work/neovim/neovim/src/nvim/normal.c:6246:5
    =     13 0x563e99bb4069 in normal_execute /home/runner/work/neovim/neovim/src/nvim/normal.c:1227:3
    =     14 0x563e9a1dcb8d in state_enter /home/runner/work/neovim/neovim/src/nvim/state.c:97:26
    =     15 0x563e99b9673c in normal_enter /home/runner/work/neovim/neovim/src/nvim/normal.c:516:3
    =     16 0x563e98c82e0d in main /home/runner/work/neovim/neovim/src/nvim/main.c:643:3
    =     17 0x7fcb81429d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 229b7dc509053fe4df5e29e8629911f0c3bc66dd)
